cmd2.py

```diff
@@ -2173,36 +2072,43 @@ def complete(self, text, state):
-                    raw_completion_token = raw_tokens[-1]
-
-                    # Add an opening quote if needed
-                    if self._handle_completion_token_quote(raw_completion_token):
-                        # An opening quote was added and the screen was updated. Return no results.
-                        self.completion_matches = []
-                        return None
-
-                    if text_to_remove or shortcut_to_restore:
-                        # If self.display_matches is empty, then set it to self.completion_matches
-                        # before we alter them. That way the suggestions will reflect how we parsed
-                        # the token being completed and not how readline did.
-                        if len(self.display_matches) == 0:
-                            self.display_matches = copy.copy(self.completion_matches)
-
-                        # Check if we need to remove text from the beginning of tab completions
-                        if text_to_remove:
-                            self.completion_matches = \
-                                [m.replace(text_to_remove, '', 1) for m in self.completion_matches]
-
-                        # Check if we need to restore a shortcut in the tab completions
-                        # so it doesn't get erased from the command line
-                        if shortcut_to_restore:
-                            self.completion_matches = \
-                                [shortcut_to_restore + match for match in self.completion_matches]
-
-                    # If the token being completed starts with a quote then we know it has an unclosed quote
-                    if len(raw_completion_token) > 0 and raw_completion_token[0] in QUOTES:
-                        unclosed_quote = raw_completion_token[0]
+                    # If self.display_matches is empty, then set it to self.completion_matches
+                    # before we alter them. That way the suggestions will reflect how we parsed
+                    # the token being completed and not how readline did.
+                    if not self.display_matches:
+                        self.display_matches = copy.copy(self.completion_matches)
+
+                    # Check if we need to add an opening quote
+                    if not unclosed_quote:
+
+                        # Get the common prefix of all matches. This is the actual tab completion.
+                        common_prefix = os.path.commonprefix(self.completion_matches)
+
+                        # Join all matches into 1 string for ease of searching
+                        all_matches_str = ''.join(self.completion_matches)
+
+                        # If there is a tab completion that will alter the text on the command line and
+                        # any of the matches have a space, then we must add an opening quote to the matches.
+                        if common_prefix != text and ' ' in all_matches_str:
```

Comment:
Thanks for this fix.  From some manual testing, this improved the behavior significantly.  The behavior now matches my expectations about how it should work.